### PR TITLE
Extracting features

### PR DESCRIPTION
**Pull request by sree-varma**

### CODE REVIEW

File: jet_images_ml.py [hunk 1]
```diff
@@ -116,170 +72,202 @@ def flatten_layer(layer):
...
- 
-        
-        #print (av_train_acc,av_validate_acc,av_val_loss)
-        if i % int(data.train.num_examples/batch_size) == 0: 
-            
-            a=session.run(merged_summary,feed_dict_train)
-            writer.add_summary(a,i)
 
+        # training step
+        _, train_acc, train_loss = session.run([optimizer, accuracy, cost], feed_dict=feed_dict_train)
```
Comment (sree-varma) [inline jet_images_ml.py:238]: To extract the features, I added the following

_, train_acc, train_loss,classifier_features = session.run([optimizer, accuracy, cost,sigmoid], feed_dict=feed_dict_train)
classifier.append(classifier_features)
and at the end of the loop (above 270),

features_class=np.concatenate(classifier,axis=0)
features_classifier=pd.DataFrame(features_class)
features_classifier.to_csv("Features_Pythia_500.csv",index=False)

where classifier=[] defined before the for loop

Am I doing the right thing?

**Comment by rbrthogan** [comment]
Ok there seems to be some funky stuff going on here

```
fc1 = fc_layer(input=layer_flat,num_inputs=num_features,num_outputs=fc_size, name="fc1")
sigmoid = tf.nn.sigmoid(fc1)
tf.summary.histogram("fc1/sigmoid", sigmoid)
fc2 = fc_layer(input=sigmoid,num_inputs=fc_size,num_outputs=num_classes, name="fc2")
```

So what we want is the have multiple layers of convolutions and fully connected layers, each activated with ReLU activations. Then, in the final layer we want to have 2 outputs (quark or gluon) and use the softmax activation (https://en.wikipedia.org/wiki/Softmax_function) (not sigmoid). More on that below. Here though you are using the wrong activation. We should instead have 

```
fc1 = fc_layer(input=layer_flat,num_inputs=num_features,num_outputs=fc_size, name="fc1")
activated_fc1 = tf.nn.relu(fc1)
fc2 = fc_layer(input=activated_fc1, num_inputs=fc_size, num_outputs=num_classes, name="fc2")
```
We don't apply any activation to fc2 yet. This is because the loss functions happen to use the logits (pre-activated outputs) for stability reasons (basically it reformulates the activation + loss function formula to avoid numerical under/overflow). Read more, https://www.tensorflow.org/api_docs/python/tf/nn/softmax_cross_entropy_with_logits.

So we need to change 
```
cross_entropy = tf.nn.sigmoid_cross_entropy_with_logits(logits=fc2, labels=y_true)
```
to 
```
cross_entropy = tf.nn.softmax_cross_entropy_with_logits(logits=fc2, labels=y_true)
```
We want to change the loss function to use softmax not sigmoid. This is because the classes are mutual exclusive and therefore the probabilities should add up to one. The sigmoid activation applied to each output could result in both probabilities being 1 or both being 0. You get around this for your metrics because you take the argmax and because for you data the labels are always (1, 0) or (0, 1) so it seems to be able to learn. You could code would not break if for example the label (0, 0) or (1, 1) however i.e. "neither quark nor gluon" or "quark and gluon" but this of course makes no sense. We should therefore enforce this in the code.

To get y_pred then, we can take argmax, but it can be useful to also get the raw probabilities first i.e.

```
y_prob = tf.nn.softmax(fc_2)
y_pred = tf.argmax(y_prob, dimension=1)
```
With y_prob we can see when it very sure of one class vs another and very it is roughly split between the two.

Now, what about the features?

```
_, train_acc, train_loss,classifier_features = session.run([optimizer, accuracy, cost,sigmoid], feed_dict=feed_dict_train)
```
should be 

```
_, train_acc, train_loss,classifier_features = session.run([optimizer, accuracy, cost, activated_fc1], feed_dict=feed_dict_train)
```

Make sense?

I have to run now, but will comment more later and reply to your email. You are also not using pull request properly. You shouldn't have to tell me what you added, I should be able to just see it. Maybe watch a youtube video on how to do pull requests in github. I'll explain more later
